src/core/d-stats.c

```diff
@@ -0,0 +1,460 @@
+                    Init_Word(ARR_AT(a, IDX_STATS_SYMBOL), f->label);
+                }
+                Init_Integer(
+                    ARR_AT(a, IDX_STATS_NUMCALLS),
+                    VAL_INT64(ARR_AT(a, IDX_STATS_NUMCALLS)) + 1
+                );
+            }
+            else if (NOT(IS_ERROR(stats))) {
+                //
+                // The user might muck with the MAP! so we put an ERROR! in
+                // to signal something went wrong, parameterized with the
+                // invalid value...as long as it isn't already an error.
+                //
+                Init_Error(stats, Error_Invalid_Arg_Raw(stats));
+            }
+        }
+
+        // Not clear if there's any statistical reason to process the r result
+        // here, but leave the scaffold in case there is.
+        //
+        switch (r) {
+        case R_FALSE:
+        r_false:
+            break;
+
+        case R_TRUE:
+        r_true:
+            break;
+
+        case R_VOID:
+        r_void:
+            break;
+
+        case R_BLANK:
+            break;
+
+        case R_BAR:
+        r_bar:
+            break;
+
+        case R_OUT:
+        r_out:
+            break;
+
+        case R_OUT_UNEVALUATED: // returned by QUOTE and SEMIQUOTE
+            goto r_out;
+
+        case R_OUT_IS_THROWN: {
+            break; }
+
+        case R_OUT_TRUE_IF_WRITTEN:
+            if (IS_END(f->out))
+                goto r_true;
+            else
+                goto r_false;
+            break;
+
+        case R_OUT_VOID_IF_UNWRITTEN:
+            if (IS_END(f->out))
+                goto r_void;
+            else
+                goto r_out;
+            break;
+
+        case R_OUT_VOID_IF_UNWRITTEN_TRUTHIFY:
+            if (IS_END(f->out))
+                goto r_void;
+            else if (IS_VOID(f->out) || IS_FALSEY(f->out))
+                goto r_bar;
+            else
+                goto r_out;
+            break;
+
+        case R_REDO_CHECKED:
+            assert(FALSE); // shouldn't be possible for final phase
+            break;
+
+        case R_REDO_UNCHECKED:
+            assert(FALSE); // shouldn't be possible for final phase
+            break;
+
+        case R_REEVALUATE_CELL:
+            break;
+
+        case R_REEVALUATE_CELL_ONLY:
+            break;
+
+        case R_UNHANDLED: // internal use only, shouldn't be returned
+            assert(FALSE);
+
+        default:
+            assert(FALSE);
+        }
+    }
+
+    return r;
+}
+
+
+//
+//  metrics: native [
+//
+//  {Track function calls and inclusive timings for those calls.}
+//
+//      return: [map!]
+//      mode [logic!]
+//          {Whether metrics should be on or off.}
+//  ]
+//
+REBNATIVE(metrics)
+{
+    INCLUDE_PARAMS_OF_METRICS;
+
+    REBVAL *mode = ARG(mode);
+
+    Check_Security(Canon(SYM_DEBUG), POL_READ, 0);
+
+    if (VAL_LOGIC(mode)) {
+        //PG_Do = &Do_Core_Measured;
+        PG_Apply = &Apply_Core_Measured;
+    }
```

Comment:
Why only the PG_Apply needs to be specialized? I see Do_Core_Measured doing nothing about the measurements, but wouldn't some evaluations miss the statistics by just collecting in Apply_Core_Measured?

Reply:
If you want to do things like count how many SET-WORD!s are executed and things like that, yes.  Although it's hard to say what exactly is happening if you only see the position before and after each instruction.  **x: i-quote-my-left y** might look like it's running a SET-WORD! assignment, but never assign anything...for instance.  So the hooks as written currently require a bit of mimicry of the evaluator if you want certain kinds of real answers.

But what I started out with here is just counting function calls.  The only place that needs to be hooked is in the apply since all function calls go through there.  I wanted to do some timing information, but it gets a bit tricky.  Not only do you need a stack of start times in the map (think about recursions) but you need to have some way of subtracting out the time for the metrics work itself.

These are just the beginnings of trying to figure out how these hooks will work.  I think there will be a generic userspace hook which will look something like:

     sees-all-calls: func [frame [frame!]] [
         print "called before every function that is not called from me"
         print "(so these prints would not get hooked, for instance)"
         print ["Frame (" label-of frame ") contents:"]
         for-each [key val] frame [
             print [mold :key "->" mold :val]
         ]

         result: do frame ;-- runs code, re-engages the hook

         print ["Result of call is:" mold :result]
         return result
     ]

     hijack function! :sees-all-calls ;-- or some other syntax

*(This technique is also how I'm thinking of accomplishing a more complex form of ADAPT, which can run code before a function executes but then process the result...like an ADAPT plus a CHAIN where it can have some state that is shared between the "preamble" and the "afterword".)*

So what I'm imagining is, that there could be simplified native forms of things like the metrics or trace hook, which you would be able to use during bootstrap.  Yet although written in C they'd still be FUNCTION!s styled like the above...we'd be the only ones who'd care about or use those, and release builds probably wouldn't have those natives.  Then, the more complex hooks for tracing or metrics would be written in userspace, that you could use when fully booted.